src/lib/communication/DY.Lib.Communication.RequestResponse.Invariants.fst

```diff
@@ -0,0 +1,211 @@
+  has_aead_predicate aead_crypto_predicates_communication_layer_and_tag
+
+
+(*** State Predicates ***)
+
+#push-options "--ifuel 2 --z3rlimit 25"
+let state_predicates_communication_layer {|crypto_invariants|}: local_state_predicate communication_states = {
+  pred = (fun tr prin sess_id st -> 
+    match st with
+    | ClientSendRequest {server; payload; key} -> (
+      let client = prin in
+      is_knowable_by (comm_label client server) tr payload /\
+      is_secret (comm_label client server) tr key /\
+      key `has_usage tr` (AeadKey comm_layer_aead_tag empty)
+    )
+    | ServerReceiveRequest {payload; key} -> (
+      let server = prin in
+      is_knowable_by (principal_label server) tr key /\
+      is_knowable_by (get_label tr key) tr payload /\
+      (
+        key `has_usage tr` (AeadKey comm_layer_aead_tag empty) \/
+        is_publishable tr key
+      )
+    )
+    | ClientReceiveResponse {server; payload; key} -> (
+      let client = prin in
+      is_knowable_by (comm_label client server) tr payload /\
+      is_secret (comm_label client server) tr key
+    )
+  );
+  pred_later = (fun tr1 tr2 prin sess_id state -> ());
+  pred_knowable = (fun tr prin sess_id state -> ());
+}
+#pop-options
+
+val state_predicates_communication_layer_and_tag: 
+  {|crypto_invariants|} ->
+  dtuple2 string local_bytes_state_predicate
+let state_predicates_communication_layer_and_tag #cinvs =
+  (|local_state_communication_layer_session.tag, local_state_predicate_to_local_bytes_state_predicate (state_predicates_communication_layer #cinvs)|)
+
+val has_communication_layer_state_predicates:
+  {|protocol_invariants|} ->
+  prop
+let has_communication_layer_state_predicates #invs =
+  has_local_state_predicate state_predicates_communication_layer
+
+
+(*** Event Predicates ***)
+
+noeq
+type comm_reqres_higher_layer_event_preds (a:Type) {| parseable_serializeable bytes a |} = {
+  send_request: tr:trace -> client:principal -> server:principal -> payload:a -> key_label:label -> prop;
+  send_request_later:
+    tr1:trace -> tr2:trace ->
+    client:principal -> server:principal -> payload:a -> key_label:label ->
+    Lemma
+    (requires
+      send_request tr1 client server payload key_label /\
+      bytes_well_formed tr1 (serialize a payload) /\
+      tr1 <$ tr2
+    )
+    (ensures
+      send_request tr2 client server payload key_label
+    )
+  ;
+  send_response: tr:trace -> server:principal -> payload:a -> prop;
+  send_response_later:
+    tr1:trace -> tr2:trace ->
+    server:principal -> payload:a ->
+    Lemma
+    (requires
+      send_response tr1 server payload /\
+      bytes_well_formed tr1 (serialize a payload) /\
+      tr1 <$ tr2
+    )
+    (ensures
+      send_response tr2 server payload
+    )
+}
+
+let default_comm_reqres_higher_layer_event_preds (a:Type) {| parseable_serializeable bytes a |} : comm_reqres_higher_layer_event_preds a = {
+  send_request = (fun tr client server payload key_label -> True);
```

Comment:
It's maybe an interesting question whether these default predicates (and default predicates elsewhere) should be constant `True` or constant `False` --- either is, I think, equally useless most of the time (since neither allows you to prove very much), but maybe there are some edge cases where it matters that I haven't thought of.

Reply:
Discussed offline --- this does indeed prevent the functions from being used by default, but means that the user has to explicitly enable the ones that are desired (and prevents them from inadvertently enabling everything at once --- see also the crypto predicates).